R/nzchar_linter.R

```diff
@@ -0,0 +1,103 @@
+#' Require usage of nzchar where appropriate
+#'
+#' [nzchar()] efficiently determines which of a vector of strings are empty
+#'   (i.e., are `""`). It should in most cases be used instead of
+#'   constructions like `string == ""` or `nchar(string) == 0`.
+#'
+#' One crucial difference is in the default handling of `NA_character_`, i.e.,
+#'   missing strings. `nzchar(NA_character_)` is `TRUE`, while `NA_character_ == ""`
+#'   and `nchar(NA_character_) == 0` are both `NA`.
+#'
+#' @evalRd rd_tags("nzchar_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+nzchar_linter <- function() {
+  # use string-length to capture both "" and ''
+  # if (any(x == "")) is not treated like it's part of if(), but
+  #   any(if (x == "") y else z) _is_ treated so. this condition looks for the
+  #   expr to be inside a call that's _not_ above an IF/WHILE.
+  comparison_xpath <- "
+  //STR_CONST[string-length(text()) = 2]
+    /parent::expr
+    /parent::expr[
+      (EQ or NE or LT or GT or LE or GE)
+      and (
+        not(ancestor-or-self::expr[
+          preceding-sibling::IF
+          or preceding-sibling::WHILE
+        ])
+        or ancestor-or-self::expr[
+          (
+            preceding-sibling::expr[SYMBOL_FUNCTION_CALL]
+            or preceding-sibling::OP-LEFT-BRACKET
+          ) and not(
+            descendant-or-self::expr[IF or WHILE]
+          )
+        ]
+      )
+    ]
+  "
+
+  # nchar(., type="width") not strictly compatible with nzchar
+  # unsure allowNA compatible, so allow it just in case (see TODO in tests)
+  nchar_xpath <- "
+  //SYMBOL_FUNCTION_CALL[text() = 'nchar']
+    /parent::expr
+    /parent::expr
+    /parent::expr[
+      (EQ or NE or LT or GT or LE or GE)
```

Comment:
Of these, only `>`, `==`, `!=` and `<=` (which is `==` in disguise) make sense. `<` is always false, `>=` is always true.

Reply:
Right... feels fine to include those anyway. `x < ""` is a code smell I guess, seems OK to lint it here. Maybe with different lint metadata?

Reply:
This one is `nchar(x) < 0` // `nchar(x) >= 0`

Reply:
😅 guessed wrong from the preview. Still feels harmless to include (and I can switch to use `infix_metadata`), WDYT?